private/buf/buflsp/buflsp.go

```diff
@@ -0,0 +1,634 @@
+		if ref := s.findReferenceAt(ctx, entry, sybmolScope, pos); ref != nil {
+			options := make(completionOptions)
+			s.findRefCompletions(ref, options)
+			return completionsToCompletionList(options), nil
+		}
+	}
+
+	// Fallback on prefix-based completions.
+	options := s.findPrefixCompletions(ctx, entry, entry.findScope(pos), entry.codeAt(pos))
+	return completionsToCompletionList(options), nil
+}
+
+func (s *server) Definition(ctx context.Context, params *protocol.DefinitionParams) ([]protocol.Location /* Definition | DefinitionLink[] | null */, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	entry, ok := s.fileCache[params.TextDocument.URI.Filename()]
+	if !ok {
+		return nil, fmt.Errorf("unknown file: %s", params.TextDocument.URI)
+	}
+	pos := entry.getSourcePos(params.Position)
+
+	var result []protocol.Location
+	if importEntry := entry.findImportEntry(pos); importEntry != nil {
+		if importEntry.docURI == "" {
+			return nil, nil
+		}
+		result = []protocol.Location{{URI: importEntry.docURI}}
+	} else {
+		result = s.findReferencedDefLoc(ctx, entry, pos)
+	}
+
+	// Make sure all the files exist.
+	for _, loc := range result {
+		if strings.HasPrefix(loc.URI.Filename(), s.container.CacheDirPath()) {
+			// This is a temporary file, make sure it exists.
+			localPath := loc.URI.Filename()
+			if _, err := os.Stat(localPath); err != nil {
+				tmpEntry, ok := s.fileCache[loc.URI.Filename()]
+				if !ok {
+					return nil, fmt.Errorf("unknown file: %s", loc.URI)
+				}
+
+				// Create the file.
+				if err := os.MkdirAll(path.Dir(localPath), 0755); err != nil {
+					return nil, err
+				}
+				tmpFile, err := os.OpenFile(localPath, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0444)
+				if err != nil {
+					return nil, err
+				}
+				defer tmpFile.Close()
+				if _, err := tmpFile.WriteString(tmpEntry.document.Text); err != nil {
+					return nil, err
+				}
+			}
+		}
+	}
+
+	return result, nil
+}
+
+func (s *server) Hover(ctx context.Context, params *protocol.HoverParams) (*protocol.Hover, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	entry, ok := s.fileCache[params.TextDocument.URI.Filename()]
+	if !ok {
+		return nil, fmt.Errorf("unknown file: %s", params.TextDocument.URI)
+	}
+	pos := entry.getSourcePos(params.Position)
+
+	symbols := s.findReferencedSymbols(ctx, entry, pos)
+	if len(symbols) == 0 {
+		return nil, nil
+	}
+	symbol := symbols[0]
+	refEntry, ok := s.fileCache[symbol.file.Filename()]
+	if !ok {
+		return nil, fmt.Errorf("unknown file: %s", symbol.file)
+	}
+
+	codeData, err := refEntry.genNodeSignature(symbol.node)
+	if err != nil {
+		return nil, err
+	}
+
+	return &protocol.Hover{
+		Contents: protocol.MarkupContent{
+			Kind:  protocol.Markdown,
+			Value: fmt.Sprintf("### %s\n```proto\n%s\n```", strings.Join(symbol.name(), "."), codeData),
+		},
+	}, nil
+}
+
+func (s *server) resolveImport(ctx context.Context, workspace bufmodule.ModuleSet, bucket bufmodule.ModuleReadBucket, path string) (*fileEntry, error) {
```

Comment:
Are all callers aware that we may have a nil error and nil value? Generally I'd expect one of the two to be non-nil with functions like this? This at least needs to be strongly documented.

Reply:
`resolveImport` no longer returns a nil error and a nil value, so I think this is safe to mark resolved.